src/der.rs

```diff
@@ -43,49 +46,49 @@ pub enum Tag {
 pub fn expect_tag_and_get_value<'a>(input: &mut untrusted::Reader<'a>,
                                     tag: Tag)
                                     -> Result<untrusted::Input<'a>,
-                                              error::Unspecified> {
+                                        ASN1Error> {
     let (actual_tag, inner) = try!(read_tag_and_get_value(input));
     if (tag as usize) != (actual_tag as usize) {
-        return Err(error::Unspecified);
+        return Err(ASN1Error);
     }
     Ok(inner)
 }
 
 pub fn read_tag_and_get_value<'a>(input: &mut untrusted::Reader<'a>)
                                   -> Result<(u8, untrusted::Input<'a>),
-                                            error::Unspecified> {
-    let tag = try!(input.read_byte());
+                                      ASN1Error> {
+    let tag = try!(read_expected_byte(input));
     if (tag & 0x1F) == 0x1F {
-        return Err(error::Unspecified); // High tag number form is not allowed.
+        return Err(ASN1Error); // High tag number form is not allowed.
     }
 
     // If the high order bit of the first byte is set to zero then the length
     // is encoded in the seven remaining bits of that byte. Otherwise, those
     // seven bits represent the number of bytes used to encode the length.
-    let length = match try!(input.read_byte()) {
+    let length = match try!(read_expected_byte(input)) {
         n if (n & 0x80) == 0 => n as usize,
         0x81 => {
-            let second_byte = try!(input.read_byte());
+            let second_byte = try!(read_expected_byte(input));
             if second_byte < 128 {
-                return Err(error::Unspecified); // Not the canonical encoding.
+                return Err(ASN1Error); // Not the canonical encoding.
             }
             second_byte as usize
         },
         0x82 => {
-            let second_byte = try!(input.read_byte()) as usize;
-            let third_byte = try!(input.read_byte()) as usize;
+            let second_byte = try!(read_expected_byte(input)) as usize;
+            let third_byte = try!(read_expected_byte(input)) as usize;
             let combined = (second_byte << 8) | third_byte;
             if combined < 256 {
-                return Err(error::Unspecified); // Not the canonical encoding.
+                return Err(ASN1Error); // Not the canonical encoding.
             }
             combined
         },
         _ => {
-            return Err(error::Unspecified); // We don't support longer lengths.
+            return Err(ASN1Error); // We don't support longer lengths.
         },
     };
 
-    let inner = try!(input.skip_and_get_input(length));
+    let inner = try!(input.skip_and_get_input(length).map_err(|_| ASN1Error));
```

Comment:
There was only one usage of skip_and_get_input so I just mapped the error inline rather than with a wrapper function.